Penumbra/UI/SharedTagManager.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Dalamud.Interface;
+using ImGuiNET;
+using OtterGui;
+using OtterGui.Raii;
+using OtterGui.Widgets;
+using Penumbra.Mods;
+using Penumbra.UI.Classes;
+
+namespace Penumbra.UI;
+public sealed class SharedTagManager
+{
+    private static uint _tagButtonAddColor = ColorId.SharedTagAdd.Value();
+    private static uint _tagButtonRemoveColor = ColorId.SharedTagRemove.Value();
+
+    private static float _minTagButtonWidth = 15;
+
+    private const string PopupContext = "SharedTagsPopup";
+    private bool _isPopupOpen = false;
+
+
+    public IReadOnlyList<string> SharedTags { get; internal set; } = Array.Empty<string>();
+
+    public SharedTagManager()
+    {
+    }
+
+    public void ChangeSharedTag(int tagIdx, string tag)
+    {
+        if (tagIdx < 0 || tagIdx > SharedTags.Count)
+            return;
+
+        if (tagIdx == SharedTags.Count) // Adding a new tag
+        {
+            SharedTags = SharedTags.Append(tag).Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+        else // Editing an existing tag
+        {
+            var tmpTags = SharedTags.ToArray();
+            tmpTags[tagIdx] = tag;
+            SharedTags = tmpTags.Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
+        }
+    }
+
+    public string DrawAddFromSharedTags(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var tagToAdd = "";
+        if (ImGuiUtil.DrawDisabledButton(FontAwesomeIcon.Tags.ToIconString(), new Vector2(ImGui.GetFrameHeight()), "Add Shared Tag... (Right-click to close popup)",
+            false, true) || _isPopupOpen)
+            return DrawSharedTagsPopup(localTags, modTags, editLocal);
+
+
+        return tagToAdd;
+    }
+
+    private string DrawSharedTagsPopup(IReadOnlyCollection<string> localTags, IReadOnlyCollection<string> modTags, bool editLocal)
+    {
+        var selected = "";
+        if (!ImGui.IsPopupOpen(PopupContext))
+        {
+            ImGui.OpenPopup(PopupContext);
+            _isPopupOpen = true;
+        }
+
+        var display = ImGui.GetIO().DisplaySize;
+        var height = Math.Min(display.Y / 4, 10 * ImGui.GetFrameHeightWithSpacing());
+        var width = display.X / 6;
+        var size = new Vector2(width, height);
+        ImGui.SetNextWindowSize(size);
+        using var popup = ImRaii.Popup(PopupContext);
+        if (!popup)
+            return selected;
+
+        ImGui.Text("Shared Tags");
+        ImGuiUtil.HoverTooltip("Right-click to close popup");
+        ImGui.Separator();
+
+        foreach (var tag in SharedTags)
+        {
+            if (DrawColoredButton(localTags, modTags, tag, editLocal))
+            {
+                selected = tag;
+                return selected;
```

Comment:
Early returns when drawing stuff will cause wonky, ugly flickering for a frame in ImGui. Recommend always drawing everything without early returns.